.ycm_extra_conf.py

```diff
@@ -1,67 +1,179 @@
+else:
+    database = None
+
+SOURCE_EXTENSIONS = [".C", ".cpp", ".cxx", ".cc", ".c", ".m", ".mm"]
 
 
 def MakeRelativePathsInFlagsAbsolute(flags, working_directory):
-  if not working_directory:
-    return flags
-  new_flags = []
-  make_next_absolute = False
-  path_flags = [ '-isystem', '-I', '-iquote', '--sysroot=' ]
-  for flag in flags:
-    new_flag = flag
-
-    if make_next_absolute:
-      make_next_absolute = False
-      if not flag.startswith('/'):
-        new_flag = os.path.join(working_directory, flag)
-
-    for path_flag in path_flags:
-      if flag == path_flag:
-        make_next_absolute = True
-        break
-
-      if flag.startswith(path_flag):
-        path = flag[ len(path_flag): ]
-        new_flag = path_flag + os.path.join(working_directory, path)
-        break
-
-    if new_flag:
-      new_flags.append(new_flag)
-  return new_flags
-
-
-def FlagsForFile(filename):
-  relative_to = DirectoryOfThisScript()
-
-  return {
-    'flags': MakeRelativePathsInFlagsAbsolute(flags, relative_to),
-    'do_cache': True
-  }
-
-flags = Flatten(map(GetFlagsFromMakefile, [
-  'AM_CPPFLAGS',
-  'CPPFLAGS',
-  'AM_CFLAGS',
-  'CFLAGS',
-]))
-
-# these flags cause crashes in libclang, so remove them
-flags.remove('-Wlogical-op')
-flags.remove('-Wsuggest-attribute=noreturn')
-flags.remove('-Wdate-time')
-
-# vim: set et ts=2 sw=2:
+    if not working_directory:
+        return list(flags)
+    new_flags = []
+    make_next_absolute = False
+    path_flags = ["-isystem", "-I", "-iquote", "--sysroot="]
+    for flag in flags:
+        new_flag = flag
+
+        if make_next_absolute:
+            make_next_absolute = False
+            if not flag.startswith("/"):
+                new_flag = os.path.join(working_directory, flag)
+
+        for path_flag in path_flags:
+            if flag == path_flag:
+                make_next_absolute = True
+                break
+
+            if flag.startswith(path_flag):
+                path = flag[len(path_flag):]
+                new_flag = path_flag + os.path.join(working_directory, path)
+                break
+
+        if new_flag:
+            new_flags.append(new_flag)
+    return new_flags
+
+
+def IsHeaderFile(filename):
+    extension = os.path.splitext(filename)[1]
+    return extension in [".H", ".h", ".hxx", ".hpp", ".hh"]
+
+
+def GetCompilationInfoForFile(filename):
+    # The compilation_commands.json file generated by CMake does not have
+    # entries for header files. So we do our best by asking the db for flags for
+    # a corresponding source file, if any. If one exists, the flags for that
+    # file should be good enough.
+    if IsHeaderFile(filename):
+        basename = os.path.splitext(filename)[0]
+        for extension in SOURCE_EXTENSIONS:
+            replacement_file = basename + extension
+            if os.path.exists(replacement_file):
+                compilation_info = \
+                    database.GetCompilationInfoForFile(replacement_file)
```

Comment:
This will crash if `database is None`. meson produces `build/compile_commands.json` so it'll work most of the time, but not when invoked before `meson` is invoked.

Reply:
Thank you for the detailed feedback, I have addressed the issues you mentioned in the existing branch, please consider merge it back to the master if there is no other issue.